# Exercice 1

Ecrire en langage python une fonction *toBase2(n)* qui a pour argument un entier positif $n$
et qui renvoie la liste des nombres $(a_k )_ {k=0..p}$ telle que 
$$n = \sum_{k=0}^{p} a_k 2^k$$ avec $a_k\in \{0,1\}$



In [1]:
def base2(n):
    res = []
    while n > 0:
        if n % 2 == 0:
            res.append(0)
        else :
            res.append(1)
        n //= 2
    return res

In [2]:
for k in range(8):
    print(base2(k))

[]
[1]
[0, 1]
[1, 1]
[0, 0, 1]
[1, 0, 1]
[0, 1, 1]
[1, 1, 1]


On constate que la fonction ne renvoie rien lorsque l'argument est négatif.
On ajoute une condition initiale pour l'argument négatif ou nul en entrée

In [3]:
def base2(n):
    res = []
    if n<= 0:
        return [0]
    while n > 0:
        if n%2 ==0:
            res.append(0)
        else:
            res.append(1)
        n //= 2
    return res

In [4]:
for k in range(8):
    print(base2(k))

[0]
[1]
[0, 1]
[1, 1]
[0, 0, 1]
[1, 0, 1]
[0, 1, 1]
[1, 1, 1]


In [5]:
l=1
base2(l)
print(l)

1


Ecrire en langage python une fonction *toBase3(n)* qui a pour argument un entier positif $n$ et qui renvoie son écriture en base 3 sous forme d’une liste.
Par exemple pour 

$$178 = 2\times 3^4 + 0\times3^3 + 1\times 3^2 + 2\times 3^1 + 1\times 3^0$$

Le programme renvoie la liste d’entiers $[1, 2, 1, 0, 2]$.


In [5]:
def toBase3(n):
    res = []
    if n<= 0:
        return [0]
    while n>0 :
        k = n%3
        res.append(k)
        n //= 3
    return res

In [6]:
toBase3(178)

[1, 2, 1, 0, 2]

In [7]:
for k in range(3**2):
    print(toBase3(k))
    

[0]
[1]
[2]
[0, 1]
[1, 1]
[2, 1]
[0, 2]
[1, 2]
[2, 2]


Ecrire en langage python une fonction *toBase(n,b)* qui convertit le nombre entier $n$ en base
$b$ ($b$ étant un entier strictement positif).

In [28]:
def toBase(n,b):
    res = []
    if n<= 0:
        return [0]
    while n>0:
        k = n % b
        #res.append(k)
        res = [k] + res
        n //= b
        # res = res[::-1]
    return res

In [29]:
for k in range(4**2):
    print(toBase(k,4))

[0]
[1]
[2]
[3]
[1, 0]
[1, 1]
[1, 2]
[1, 3]
[2, 0]
[2, 1]
[2, 2]
[2, 3]
[3, 0]
[3, 1]
[3, 2]
[3, 3]


# Exercice 2 Nombres premiers jumeaux

Ecrire en langage python une fonction *jumeaux* qui renvoie une liste contenant les couples de
nombres premiers jumeaux inférieurs (strictement) à un entier positif n donné comme argument.

**Définitions (rappels)**

Un *nombre premier* est un entier naturel qui admet exactement deux diviseurs : $1$ et lui-même.
Ainsi $1$ n'est pas considéré comme premier.

Deux nombres premiers $p$, $q$ forment un *couple de nombres premiers jumeaux* $(p,q)$ si
$$q=p+2$$

<font color=red>**Solution** </font>
Plusieurs méthodes sont envisageables :

1. On peut par exemple rechercher la liste des nombres premiers inférieurs à $n$ (voire $n-2$) puis calculer les distances entre deux nombres successifs afin de détecter les nombres premiers jumeaux.

2. On peut également parcourir tous les entiers $k$ inférieurs $n$ par une boucle "FOR" et faire un test de primalité sur chacun. Sitôt que l'un est premier, on teste la primalité de $k+2$ (ou de k-2). Si deux successifs sont premiers, on ajoute le couple à une liste "resultats" qui doit être renvoyée par la fonction.

On choisit d'appliquer la seconde méthode qui évite de construire la liste de nombres premiers car il ne semble pas nécessaire de retenir tous les nombres premiers, mais seulement ceux qui participent à un couple de jumeaux.
A ce stade, voici donc la *trame* de l'algorithme:

```python
  nbCouples <-0 # nombre de couples trouvés
  listeJumeaux <-[] # liste Vide
    
  POUR k allant de 1 à nMax
      SI (k est premier)
           SI (k+2 est premier)
                ajouter le couple [k,k+2] à listeJumeaux
  RETOURNER listeJumeaux
```
Il reste à définir la valeur maximale prise par l'indice $k$ dans cette boucle.
Sachant que les deux nombres du couple $[k,k+2]$ doivent être inférieurs strictement à $n$.
Il faut donc
$$ k+2<n$$
ainsi la valeur $nMax$ vaut donc $n-3$.

On constate que nous avons besoin d'une seconde fonction auxiliare pour tester la *primalité* d'un nombre entier (qui renvoie *True* si le nombre est premier et *False* sinon).
Ensuite, il parait logique de travailler avec une liste, considérée comme variable globale que l'on soit susceptible de compléter pour y ajouter des nombres premiers.

**Etape 1 : test de primalité**
Pour tester la primalité d'un nombre on va tester tous ses potentiels diviseurs.

Exemple : pour savoir si $n=83$ est premier, on va calculer le reste de la division de $83$ par tous les entiers $k$ tels que $$k\times k <= 83$$
Dès que l'on trouve un reste de la division de 83 par $k$ qui est nul, on retourne *False*. Si on parcourt la boucle sans en sortir, on retourne *True*.

A quelle valeur minimale pour $k$ doit-on partir? Le plus petit diviseur potentiel permettant de savoir si un nombre est premier est nombre $2$.
Attention, si $n=2$, il sera évidemment divisible par $2$ (le nombre $2$ n'est divisible que par 1 et par lui-même, il est donc premier). Toutefois, ce cas ne sera pas testé car "$2 \times 2\le 2$" est fausse et donc l'entrée dans la boucle ne se fait pas.

Attention: il faut bien vérifier que nos tests de diviseurs potentiels (l'ensemble des valeurs prises par la variable $k$) contient la racine carré de $n$. Exemple pour $n=36$, il faut que $k$ puisse atteindre $k=6$ dans la boucle d'exécution. Or, cela est le cas la condition d'entrée de boucle est bien $$k\times k \le n$$ et donc l'usage du test **inférieur ou égal** dans cette condition est indispensable.

Enfin, les nombres 0 et 1 ne sont pas premiers et il en va de même pour les entiers négatifs. On peut donc traiter ces cas dès l'entrée de boucle en renvoyant *False* si $n<2$.

In [30]:
def isPrime(n):
    if n < 2:
        return False
    k = 2
    while k * k <= n:
        if n % k == 0:
            return False
        k += 1
    return True


In [31]:
# Validation en affichant les nombres premiers 
# inférieurs strictement à 25
for k in range(25):
    if isPrime(k):
        print(k)

2
3
5
7
11
13
17
19
23


**Etape 2 : liste des couples de jumeaux**
On définit la fonction jumeaux qui, à partir d'un argument entier $n$, renvoie les couples de jumeaux strictement inférieurs à $n$.

In [6]:
def jumeaux(n):
    listeJumeaux = [] #initialisation de la liste
    for k in range(n - 2): # la valeur maximale prise par k est n-3
                         # si bien que les deux valeurs du couple
                         # [k,k+2] sont strictement inférieures à n
        if isPrime(k):
            if isPrime(k + 2):
                listeJumeaux.append([k, k + 2])
    return listeJumeaux

In [7]:
# Validation avec l'exemple de l'énoncé
jumeaux(109)

[[3, 5],
 [5, 7],
 [11, 13],
 [17, 19],
 [29, 31],
 [41, 43],
 [59, 61],
 [71, 73],
 [101, 103]]

**Critique du code proposé**

Avec cet algorithme, le test de primalité est effectué deux fois par boucle. Or, tester $k$ et $k+2$ est redondant car, deux itérations suivantes, $k$ sera augmenté de $2$ et donc on testera à nouveau la primalité du même entier.
On peut donc optimiser ce code en stockant la valeur du dernier nombre premier trouvé à chaque itération et vérifier si le suivant trouvé est égal à l'ancien augmenté de deux.

Voici l'algorithme qui illustre cette idée:
```python
nbCouples <-0 # nombre de couples trouvés
listeJumeaux <-[] # liste Vide
kOld = 2 # c'est le premier nombre premier    
POUR k allant de 1 à nMax
      SI (k est premier)
           SI (k= kOld+2)
                Ajouter [kOld,k] à listeJumeaux
           kOld=k # mise à jour à nombre premier trouvé
  RETOURNER listeJumeaux
```
Il reste à définir la valeur maximale prise par l'indice $k$ dans cette boucle.
Sachant que les deux nombres du couple $[kOld,k]$ doivent être inférieures strictement à $n$.
Il faut donc
$$ k<n$$
ainsi la valeur $nMax$ vaut donc $n-1$.

In [15]:
#Voici le code python de cet algorithme
def jumeauxBis(n):
    listeJumeaux = [] #initialisation de la liste
    kOld = 2
    for k in range(n): # la valeur maximale prise par k est n-1
                       # si bien que les deux valeurs du couple
                       # [kold, k] sont strictement inférieures à n
        if isPrime(k) :
            if (k == kOld + 2):
                listeJumeaux.append([kOld, k])
            kOld = k # mis à jour à chaque nb premier trouvé
                   # ATTENTION : à l'indentation de cette instruction
    return listeJumeaux

In [16]:
#Validation
jumeauxBis(109)

[[3, 5],
 [5, 7],
 [11, 13],
 [17, 19],
 [29, 31],
 [41, 43],
 [59, 61],
 [71, 73],
 [101, 103]]

# Exercice 2bis Nombres premiers jumeaux
Ecrire en langage python une fonction *premierJumeaux* qui, à partir d'un entier positif donné comme argument, renvoie une liste contenant les $n$ *premiers couples* de nombres premiers jumeaux.

**Définitions (rappels)**

Un *nombre premier* est un entier naturel qui admet exactement deux diviseurs : $1$ et lui-même.
Ainsi $1$ n'est pas considéré comme premier.

Deux nombres premiers $p$,$q$ forment un *couple de nombres premiers jumeaux* $(p,q)$ si
$$q=p+2$$




<font color=red>**Solution** </font>
Plusieurs méthodes sont envisageables :

1. On peut par exemple rechercher la liste des *k* plus petits *nombres premiers* puis calculer les distances successives entre eux. Par contre, on ne sait pas combien de nombres premiers il faut considérer car notre objectif n'est pas de trouver des entiers mais des nombres premiers jumeaux.

2. On peut également parcourir tous les entiers et faire des tests de primalité sur chacun. Si tôt que l'un est premier, on teste la primalité de n+2. (ou de n-2). On arrête le parcours de tous les entiers dès que l'on a trouvé $n$ couples de jumeaux.

Conclusion : rechercher la liste des plus petits nombres premiers requiert a priori de parcourir tous les nombres pour tester leur primalité. Ainsi, autant chercher directement la liste de tous les nombres premiers successivement par ordre croissant. A mesure que l'on remplit cette liste, on vérifie si le nouveau nombre premier trouvé est séparé du dernier de deux unités seulement. Dans ce cas, on incrémente un compteur de couples de nombres jumeaux obtenus. La sortie de boucle se ferait donc lorsque l'on a obtenu suffisamment de couples de jumeaux.

A ce stade, voici la *trame* de l'algorithme:

```python
  nbCouples <-0 # nombre de couples trouvés
  listeJumeaux <-[] # liste Vide
    
  k=2 # nombre premier initial
  TANT QUE nbCouples < n                    
      knext <- premierSuivant(k)                     
      SI (knext==k+2)                        
          AJOUTER (k,knext) à listejumeaux          
          nbCouples <-nbCouples+1
      k <-knext #faire pointer k sur le suivant
  RETOURNER listeJumeaux
```

Nous avons naturellement introduit une fonction renvoyant le nombre premier suivant un entier donné.
Il s'agit d'une fonction qui reçoit en argument un entier et propose le nombre premier immédiatement supérieur à cet entier. Pour cela, il semble possible de parcourir tous les entiers à partir de ce point de départ et de s'arrêter dès que l'on trouve un nombre premier que l'on renvoi alors comme résultat de la fonction.

Voici la trame algorithmique de cette fonction auxiliaire:


```python
FONCTION premierSuivant(n)
    k=n+1 # on teste à partir de n+1 pour
          # renvoyer le suivant
    TANT QUE (k n'est pas premier)
              k <- k+1
    RETOURNER k
```


On constate que nous avons besoin d'une seconde fonction auxiliare pour tester la *primalité* d'un nombre entier (qui renvoie *True* si le nombre est premier et *False* sinon).
Ensuite, il parait logique de travailler avec une liste, considérée comme variable globale que l'on soit susceptible de compléter pour y ajouter des nombres premiers.

**Etape 1 : test de primalité**
Pour tester la primalité d'un nombre on va tester tout ses potentiels diviseurs.

Exemple : pour savoir si $n=83$ est premier, on va calculer le reste de la division de $83$ par tous les entiers $k$ tels que $$k\times k < 83$$
Dès que l'on trouve un reste de la division de 83 par $k$ qui est nul, on retourne *False*. Si on parcourt la boucle sans en sortir, on retourne *True*.

A quelle valeur minimale pour $k$ doit-on partir? Le plus petit diviseur potentiel permettant de savoir si un nombre est premier est nombre $2$.
Attention, si $n=2$, il sera évidemment divisible par $2$ (le nombre $2$ n'est divisible que par 1 et par lui-même, il est donc premier). Toutefois, ce cas ne sera pas testé car "$2 \times 2\le 2$" est fausse et donc l'entrée dans la boucle ne se fait pas.

Attention: il faut bien vérifier que nos tests de diviseurs potentiels (l'ensemble des valeurs prises par la variable $k$) contient la racine carré de $n$. Exemple pour $n=36$, il faut que $k$ puisse atteindre $k=6$ dans la boucle d'exécution. Or, cela est le cas la condition d'entrée de boucle est bien $$k\times k \le n$$ et donc l'usage du test **inférieur ou égal** dans cette condition est indispensable.

Enfin, les nombres 0 et 1 ne sont pas premiers et il en va de même pour les entiers négatifs. On peut donc traiter ces cas dès l'entrée de boucle en renvoyant *False* si $n<2$.

In [17]:
def isPrime(n):
    if n < 2:
        return False
    k = 2
    while k * k <= n:
        if n % k == 0:
            return False
        k += 1
    return True


In [18]:
# Validation en affichant les nb premiers 
# inférieurs strictement à 25
for k in range(25):
    if isPrime(k):
        print(k)

2
3
5
7
11
13
17
19
23


**Etape 2 : renvoyer le premier suivant**
On rédige le code python d'une fonction qui reçoit en argument un entier et qui retourne le nombre premier immédiatement supérieur à n

In [19]:
def premierSuivant (n):
    k = n + 1
    while not isPrime(k):
        k += 1
    return k

In [20]:
# Validation : affichage des 10 premiers nombres premiers
n = 0
for k in range(10):
    n = premierSuivant(n)
    print(n)

2
3
5
7
11
13
17
19
23
29


In [21]:
# On vérifie également pour une liste de valeurs
for k in range(12):
    print(k,'->',premierSuivant(k))

0 -> 2
1 -> 2
2 -> 3
3 -> 5
4 -> 5
5 -> 7
6 -> 7
7 -> 11
8 -> 11
9 -> 11
10 -> 11
11 -> 13


**Etape 3 : renvoyer la liste de nombre jumeaux**
On rédige le code python d'une fonction *jumeaux* qui, à partir d'un argument entier $n$ renvoie les $n$ premiers couples de nombres jumeaux.

In [22]:
def premierJumeaux(n):
    nbCouples = 0
    listeJumeaux = []
    k = 2 #nombre initial
    while (nbCouples < n):
        knext = premierSuivant(k)
        if (knext == k + 2):
            listeJumeaux.append([k,knext])
            nbCouples += 1
        k=knext
    return listeJumeaux

In [23]:
# La liste des n=10 premiers couples 
# de nombres de jumeaux
premierJumeaux(10)

[[3, 5],
 [5, 7],
 [11, 13],
 [17, 19],
 [29, 31],
 [41, 43],
 [59, 61],
 [71, 73],
 [101, 103],
 [107, 109]]

# Correction sujet IPT  Mines 2019 

Questions Q1 à Q10

## Q1 
Dans un programme Python on souhaite pouvoir faire appel aux fonctions log, sqrt, floor et ceil du module math (round est disponible par défaut). Ecrire des instructions permettant d’avoir accès à ces fonctions et d’afficher le logarithme népérien de 0.5.

Deux possibilités ``import *`` ou `` import as``

In [24]:
from math import *
print(log(0.5)) # logarithme népérien

-0.6931471805599453


In [32]:
import math as mth
print(mth.log(0.5)) # les fonctions sont précédés
                    # de l'alias mth

-0.6931471805599453


In [35]:
mth.log10(10)

1.0

## Q2 Fonction ``sont_proches(x, y )``

In [26]:
def sont_proches(x, y) :
    atol = 1e-5
    rtol = 1e-8
    return abs(x-y)<= atol + abs(y) * rtol    

In [37]:
3<2

False

## Q3 fonction ``mystere(x, b)``

Pour déterminer ce que renvoit ``mystere(1001,10)``, on exécute mentalement l'algorithme en Python.

- au premier appel, x = 1001 et b = 10 donc ``x < b`` est FAUX. Ainsi, la fonction renvoie 1 +`` ``mystere(1001/10,10)``
- au second appel, x = 1001/10 et b = 10. Ainsi, on a toujours ``x < b`` qui est FAUX. On ajoute encore 1 au résultat.
- au 3ème appel, x = 1001/100 et b = 10. Ainsi, on a encore ``x >= b `` donc on ajoute encore 1 au résultat.
- au 4ème appel, x = 1001/1000 et b = 10. Cette fois ``x < b`` devient VRAI. La fonction renvoie zéro. 
 Au final, la fonction ``mystere`` est ainsi appelée 4 fois et **elle renvoie la valeur 3**.


## Q4
En généralisant ce qui précéde, on peut dire que :
lors du *k-ième appel* de la fonction ``mystere``, il y a eu *k-1* additions du terme 1. L'argument initial $x$ a été divisé *k-1* fois par b. Donc, la valeur de l'argument ``x`` pour le *k-ième* appel est :
$$x_k=x/b^{k-1}$$
L'appel à cette fonction s'arrête dès que $x_k< b$ soit lorsque
$$  x/b^{k-1} < b$$
Où encore $x < b^{k}$

Soit, $\ln(x) < k \ln(b),$ c'est-à-dire pour l'itération *k* telle que  
$$k > \frac{\ln(x)}{\ln(b)}$$
Le plus petit entier $k$ vérifiant cette condition est la *partie entière inférieure* additionnée de un (prendre par exemple les cas "k>3,4" et "k>3" pour s'en persuader):
$$k_0=\left\lfloor\frac{\ln(x)}{\ln(b)} \right\rfloor +1$$
Or, à cette *k-ième* itération, la fonction additionne avec zéro. Ainsi, elle renvoie $k_0-1$. In fine, la fonction ``mystere``  renverra la valeur 
$$\left\lfloor\frac{\ln(x)}{\ln(b)} \right\rfloor$$
et sera appelée $\left\lfloor\frac{\ln(x)}{\ln(b)} \right\rfloor+1$ fois.
Note: cette valeur renvoyée s'écrit également en Python par : ``floor(log(x)/log(b))``.


## Q5 
- La variable ``x1`` est égale à $i+1$ multiplié par le pas. A la fin des 100000 itérations, elle est le résultat de 2 opérations (une addition puis une multiplication). Sa valeur numérique telle qu'elle est représentée en machine est ainsi, a priori, très proche de sa valeur exacte ($100000 \times 10^{-5}$).

- La variable ``x2`` quant à elle est, à chaque itération, calculée en additionnant le contenu de la variable ``pas``. Compte-tenu de **la précision limitée du codage des nombres en machine**, les additions successives de la variable ``x2`` se font potentiellement avec de légères erreurs d'arrondis. Ainsi, au bout de 100000 itérations, les additions successives *donnent un résultat différent* de celui que l'on obtient directement par la multiplication.

On note que le résultat obtenu est x2=0.9999999999980838, il possède 11 chiffres corrects. L'erreur relative est donc de $10^{-11}$. Sachant que la précision de codage d'un nombre flottant en Python et de $10^{-15} $ à  $10^{-16} $, on retrouve bien que l'accumulation des $100000 = 10^5$ petites erreurs de l'ordre de $10^{-16}$ donne une erreur relative de:
$$10^{-16} \times 10^{5} = 10^{-11}$$
Il est donc logique d'avoir environ 11 chiffres significatifs dans la valeur de la variable ``x2``.

In [27]:
pas = 1e-8 # pas encore plus petit
x2 = 0
for i in range (100000000): # 10^8 itérations
    x1 = ( i + 1)  * pas
    x2 = x2 + pas
print ( 'x1 : ' , x1 )
print ( 'x2 : ' , x2 )


x1 :  1.0
x2 :  1.0000000022898672


On note que 1.0000000022898672 possède 8 chiffres corrects seulement.

## Q6

Une liste de $N$ booléens aura sa taille qui augmente de 32 bits soit de 4 octets (rappel 1 octet = 8 bits) avec chaque nouveau booléen stocké dans cette liste.

Ainsi, avec 4 Go (giga octets) $= 4 \times 10^9$ octets de mémoire vive, on peut stocker
$$ N=4\times 10 ^9 / 4 = 10^9$$
valeurs booléennes.

## Q7
Le plus petit espace mémoire pour stocker un booléen est le bit (information binaire à deux états, tout comme un booléen True ou False). On peut ainsi **gagner un facteur 32**.

## Q8 Fonction ``erato_iter(N)``

Attention : cette question est plus difficile qu'elle peut sembler au premier abord. Il importe de bien vérifier *les effets de bords*, c'est-à-dire:
 - vérifier le nombre d'itérations dans les boucles,
 - l'indice des éléments minimum et maximum,
 - comprendre le lien entre le *k-ième* élément de ``liste_bool``, le nombre dont il stocke la primalité et l'indice de cet élément.
 - attention à employer rigoureusement l'instruction ``range``.
 
 Important, l'énnoncé nous aide en donnant un exemple: pour $N=4$, la fonction renvoie une liste de 4 éléments [False, True,True,False] correspondant à la primalité des nombres 1,2,3,4.
 Ainsi, le premier élément de la liste contient la primalité du nombre 1. 
 
 Donc
 **le k-ième élément de la liste, ``liste_bool[k-1]`` contient la primalité du nombre k.**

Voici les principales difficultés :
Comment initialiser une liste avec N booléens à ``True``?

In [ ]:
liste_bool = [True] * N # N concaténations de liste
# ou bien
liste_bool = [ True for k in range(N)]

Comment faire varier $i$ de 2 à $\left \lfloor \sqrt{n}\right \rfloor$ ?

In [ ]:
for i in range(2,floor(sqrt(N))+1): # CORRECT !

for i in range(2,floor(sqrt(N))): # FAUX ! 

attention il faut bien le " +1 ", sinon la valeur finale n'est pas atteinte

Comment "marquer faux tous les multiples de $i$ différents de $i$ dans list_bool? 

On pourrait être tenté d'utiliser un sciling de liste mais cela est un peu technique car l'affectation du genre:

``L[a:b:step] = False``
est impossible car le membre de gauche de l'affectation est une liste et le membre de droite est un booléen. Ce type d'affectation n'est pas légitime pour des listes en Python (Notons qu'elle le serait toutefois pour des tableaux Numpy).

On pourrait contourner cette difficulté grâce à l'astuce suivante :

In [ ]:
L[a:b:step] = [False] * len(L[a:b:step])

qui permet de générer dans le membre de droite une liste initialisée à ``False`` comportant exactement le même nombre de termes que ceux devant être initialisés.

Une méthode plus *"classique"* (et plus sûre en concours) et d'utiliser une boucle ``for`` qui parcourt les éléments depuis le premier multiple de $i$, qui n'est pas $i$ (on doit partir de $2i$) jusqu'au dernier de la liste (donc le $N$-ième doit être atteint !) en incrémentant de $i$ à chaque itération.

Rappel ``for k in range(a,b,step):``
les valeurs prises par *k* sont *a*, *a+step*, *a+2step*, ... mais la valeur *b* **n'est pas atteinte**.

In [ ]:
for k in range(2*i,N+1,i): # attention N doit être atteint
                liste_bool[k-1] = False

Attention : ``liste_bool[k-1]`` désigne bien **le *k-ième* élément** de la liste, c'est celui qui code la primalité du nombre *k*.

Voici une implémentation possible de la fonction demandée:

In [2]:
def erato_iter(N):
    if N < 1 :
        print('N doit être >=1 ')
        return
    liste_bool = [True] * N # N concaténations de liste
    liste_bool[0] = False # c'est le nombre 1
    for i in range(2,floor(sqrt(N))+1):
        if liste_bool[i-1] : # attention à l'indice
            for k in range(2*i,N+1,i):
                liste_bool[k-1] = False
    return liste_bool


Validation de la fonction sur un exemple ci-dessous.

In [ ]:
# affiche les nombres premiers inférieurs ou égaux à 25
LL=erato_iter(25) 
for k in range(len(LL)):
    if LL[k]: # LL[k] ==  primalité de k+1
        print(k+1)

## Q9

Dans la question, la complexité doit être calculée *en ordre de grandeur*, il n'est pas nécessaire de compter **à l'opération près**. On peut considérer comme **opération élémentaire** le fait de 
*marquer comme Faux une valeur de la liste ``liste_bool``*.

Pour évaluer la complexité algorithmique du crible d'Eratosthène, notons qu'il comporte une boucle qui s'effectue sur $\left \lfloor \sqrt{n}\right \rfloor$ premières valeurs du tableau itérations.

Deux cas sont possibles :

- ou bien la valeur est déjà marquée comme Faux (le nombre n'est pas premier), dans ce cas, il est ignoré.
- ou bien la valeur n'est pas marquée comme Faux (le nombre est considéré comme premier), dans ce cas on entre dans la seconde boucle.

La seconde boucle s'effectue pour chaque nombre premier $p$ inférieur à $N$. Elle constite à marquer tous ses multiples (à l'exception de lui-même) : 2p, 3p, 4p, ..., kp jusqu'à dernier inférieur ou égal à $N$: 
**il y a donc de l'ordre de $N/p$ marquages** dans le tableau pour chaque itération de cette boucle.

Ainsi, le nombre total de marquages est donc de 
$$ \sum_{p<N,\textrm{ $p$ premier}} \frac{N}{p} \approx N \ln(\ln(N))$$
La complexité algorithmique du crible est donc
$$C(n) \sim  N \ln(\ln(N))$$
Remarque : on vérifie évidemment qu'elle est supérieure à la complexité des $\left \lfloor \sqrt{n}\right \rfloor$ termes de la première boucle:
$$\lim_{n\to\infty} \sqrt{n}/C(n) =0$$.

## Q10 Complexité en fonction de la base

- En base 10, le nombre 1000 = $10^3$ s'écrit avec 4 chifres.

- En base 10, le nombre $10^k$, c'est écrit avec $k+1$ chiffres.

Ainsi, un $N=b^n$ s'écrit en base $b$ avec $n+1$ chiffres. Son nombre de chiffres en base $b$ est donné en écrivant
$$\ln(N)=n \ln(b)$$
soit $$n=\frac{\ln(N)}{\ln(b)}$$
Ainsi, en base $b$, le nombre $N$ s'écrit avec un nombre de chiffres de l'ordre de $\frac{\ln(N)}{\ln(b)}$.

En base $b$, la complexité algorithmique du crible est donc:
$$C(n)\sim N \ln(\ln(N)) \sim b^n \ln(n\ln(b)) $$
La complexité, exprimée en fonction du nombre de chiffres $n$  en base 10, est donc
$$C_{\textrm{b=10}}(n) \sim 10^n \ln(n\ln(10)) $$
